API Sideload Automation

rapheld opened the pull request:
Automate the loading of sideload API data into an Ember.Resource instance.

rapheld committed:
syntax cleanup

rapheld committed:
Adding sideload functionality

rapheld commented:
@shajith @staugaard @jamesarosen

rapheld commented:
Could someone give this a review today...only a couple dozen lines of code excluding tests.

shajith commented on dist/ember-resource.js:
```diff
@@ -1028,10 +1030,31 @@ if (typeof this === 'object') this.LRUCache = LRUCache;
+      var sideloads = this.constructor.sideloads;
+      if (!sideloads || Ember.keys(sideloads).length === 0) { return; }
+
+      for (var name in sideloads) {
```
Do a `hasOwnProperty` check here.

rapheld replied:
Good catch...added `hasOwnProperty`

shajith commented on dist/ember-resource.js:
```diff
@@ -1028,10 +1030,31 @@ if (typeof this === 'object') this.LRUCache = LRUCache;
+        var klass    = sideloads[name],
+            instance = Ember.get(this, name);
+
+        if (typeof klass === 'string') { klass = Ember.getPath(klass); }
```
There is an `isString` you can use.

rapheld replied:
Oh yeah...added `isString`

shajith commented on dist/ember-resource.js:
```diff
@@ -1068,8 +1091,8 @@ if (typeof this === 'object') this.LRUCache = LRUCache;
 
-      if(sideloads && sideloads.length !== 0) {
-        ajaxOptions.data = {include: sideloads.join(",")};
+      if(sideloads && Ember.keys(sideloads).length !== 0) {
```
Maybe cache the result  of `Ember.keys`?

rapheld replied:
good call. I cleaned this up as well as the defense clause in `updateSideloadApiData`. Much better, thank you.

shajith commented on spec/javascripts/resourceSpec.js:
```diff
@@ -17,30 +17,59 @@ describe('A Resource instance', function() {
+        food.fetch();
       });
 
-      it("should not include the sideloads in resourceURL", function() {
```
Why is this test gone?

rapheld replied:
It is covered with another test:

``` javascript
      it("should send the sideloads in AJAX fetches", function() {
        expect(Em.Resource.ajax).toHaveBeenCalledWith({
          url: "/foods/1",
          resource: food,
          operation: 'read',
          data: {include: "burger,smoothie"}
        });
      });
```

shajith replied:
No, the test you deleted says: `"should not include the sideloads in resourceURL"`

rapheld replied:
True, but the `url` ajax option is the value of `resourceURL`. I can definitely add it back in though.

`var url = this.resourceURL();`

shajith replied:
Right, we need a separate assertion about `resourceURL` not containing sideloads.

rapheld replied:
Test added back in.

shajith commented on dist/ember-resource.js:
```diff
@@ -1028,10 +1030,31 @@ if (typeof this === 'object') this.LRUCache = LRUCache;
+          Ember.set(this, name, instance);
+        }
+        var data = Ember.get(instance, 'data');
+        Ember.Resource.deepMerge(data, instance.constructor.parse(json[name]));
```
Have you tested with sideloaded `ResourceCollection`s? I am not sure if we can instantiate them this way.

rapheld replied:
Not sure I'm following you. Can you give an example?

shajith replied:
What happens when you sideload a collection of resources, that is (if that is even possible).

rapheld replied:
For now, I think the answer is, if a sideload is a collection, the parse method will manage that. I will look into automating that after.

shajith replied:
Could you test that? I don't think the technique of `deepMerge`-ing data will work with collections.

Which leads me to: why do you access `data`, or explicitly call `parse` at all? Could you not call `updateWithApiData` on the instance with the data in the sideloaded JSON?

rapheld replied:
You are correct, `deepMerge` will not work with a collection. I can test that the current logic doesn't break when an array is passed in for a sideload response. I haven't built the logic to automatically attach a `ResourceCollection`, though plan to. Does this make sense? We can discuss more that the BBQ tomorow.

So I don't call `updateWithApiData` from within `updateSideloadApiData` for two reasons. Currently, `updateWithApiData` actually contains the call to `updateSideloadApiData` which would cause in infinite loop. `updateWithApiData` also contains `beginPropertyChanges` and `endPropertyChanges` which should only be called once for the whole update process (the resource and all its sideloads). `updateSideloadApiData` is a recursive method, so this way it is completely within the `PropertyChanges` calls.

Both update methods do contain pretty much the same `deepMerge` line of code: `Ember.Resource.deepMerge(data, this.constructor.parse(json));`. I could refactor that into a separate method.

shajith replied:
Not sure I see the infinite loop: You would be calling `updateWithApiData` on a sub-resource that was sideloaded, not on the same instance, right? I.e within, say, `lunch.updateWithApiData`, you'd call `burger.updateWithApiData` (assuming lunch sideloads burger).

Also: we do call `beginPropertyChanges`/`endPropertyChanges` pairs whenever we create resources, the nesting is the expected behavior and well supported by Ember. These are also called on the sub resources.

rapheld replied:
You are absolutely right. I modified the code to use `updateWithApiData`...no infinite loop. Also, now `updateSideloadApiData` can now handle collections. Check it out and let me know what you think.

shajith commented:
Sorry I left the comments on the dist file. Maybe we should split out the dist file update into a separate version bump update?

Also: Do verify interaction with `ResourceCollection`s.

rapheld committed:
Add recursion for nested API sideloading

Allows automatic Ember Resource creation when working with with complex nested API objects

rapheld commented:
RE: `ResourceCollection`

Are you referring to a case where the sideload is a collection of, objects like users off a ticket? I will look in to this.

rapheld commented:
Also, I added a few lines of code to recurse over nested resources (such as found within account settings (e.g. `Zd.currentAccount.settings.apps.get('use')`

Here `settings` is a sideload that contains multiple resources such as `apps`

rapheld committed:
adding removed resource test

rapheld committed:
Add ability to handle sideload collection

Creates an Ember.ResourceCollection with items of the class type
specified within the sideload definition

rapheld commented:
OLD
